src/components/intent.ts

```diff
@@ -0,0 +1,850 @@
+        else:
+        FAIL (bot can't get into the room)
+        */
+
+        const deferredPromise = defer();
+
+        const mark = (room: string, state: MembershipState) => {
+            this.opts.backingStore.setMembership(room, userId, state);
+            if (state === "join") {
+                deferredPromise.resolve();
+            }
+        }
+
+        const dontJoin = this.opts.dontJoin;
+
+        try {
+            await this._ensureRegistered();
+            if (dontJoin) {
+                deferredPromise.resolve();
+                return deferredPromise.promise;
+            }
+            try {
+                await this.client.joinRoom(roomId, opts);
+                mark(roomId, "join");
+            }
+            catch (ex) {
+                if (ex.errcode !== "M_FORBIDDEN") {
+                    throw ex;
+                }
+                try {
+                    // Try bot inviting client
+                    await this.botClient.invite(roomId, userId);
+                    await this.client.joinRoom(roomId, opts);
+                    mark(roomId, "join");
+                }
+                catch (_ex) {
+                    // Try bot joining
+                    await this.botClient.joinRoom(roomId, opts)
+                    await this.botClient.invite(roomId, userId);
+                    await this.client.joinRoom(roomId, opts);
+                    mark(roomId, "join");
+                }
+            }
+        }
+        catch (ex) {
+            deferredPromise.reject(passthroughError ? ex : Error("Failed to join room"));
+        }
+
+        return deferredPromise.promise;
+    }
+
+    private async _ensureHasPowerLevelFor(roomId: string, eventType: string) {
+        if (this.opts.dontCheckPowerLevel && eventType !== "m.room.power_levels") {
+            return Promise.resolve();
+        }
+        const userId = this.client.credentials.userId;
+        const plContent = this.opts.backingStore.getPowerLevelContent(roomId);
+        let promise = Promise.resolve(plContent);
+        if (!plContent) {
+            promise = this.client.getStateEvent(roomId, "m.room.power_levels", "") as Promise<PowerLevelContent>;
+        }
+        return promise.then((eventContent) => {
+            this.opts.backingStore.setPowerLevelContent(roomId, eventContent);
+            const event = {
+                content: eventContent,
+                room_id: roomId,
+                sender: "",
+                event_id: "_",
+                state_key: "",
+                type: "m.room.power_levels"
+            }
+            const powerLevelEvent = new MatrixEvent(event);
+            // What level do we need for this event type?
+            const defaultLevel = STATE_EVENT_TYPES.includes(eventType)
+                ? event.content.state_default
+                : event.content.events_default;
+            const requiredLevel = event.content.events[eventType] || defaultLevel;
```

Comment:
If the powerLevelEvent's content did not contain a required level for the eventType and also no defaults, what should happen? Do we just assume the required power level is higher than what we have?

Reply:
I believe in theory synapse ensures that should never happen, but we *should* [follow the spec](https://matrix.org/docs/spec/client_server/r0.6.0#m-room-power-levels) which mentions what values to use if the keys do not exist.